namui-cli/src/build/build_server/code_watcher.rs

```diff
@@ -0,0 +1,167 @@
+use crate::debug_println;
+use cargo_metadata::MetadataCommand;
+use futures::{executor::block_on, lock::Mutex};
+use notify::{DebouncedEvent, INotifyWatcher, RecursiveMode, Watcher};
+use regex::Regex;
+use std::{collections::HashSet, path::PathBuf, sync::Arc, thread, time::Duration};
+use tokio::sync::oneshot::{self, Sender};
+
+struct CodeWatcherContext {
+    was_changed: bool,
+    sender: Option<Sender<()>>,
```

Comment:
It looks queue but only one loop uses `CodeWatcher`.
So It will not be array of senders, but will be only one sender as optional

Reply:
Is `only one loop` mean `only one consumer`?
If it's right, why don't you just make a callback parameter and just call it? because it's very simple and easy to debug.

Reply:
If we use callback, we will probably call callback in watcher thread.
It will definitely be simpler than now.

But there's something I'm worried about.
If we block watcher thread, fs events will accumulate.
I think it's better to consume the event as soon as possible.
This is why I process events in another thread, the watcher thread.

But if you can rebuild on the main thread while using callback, that's probably a better way.
If you know how, can you recommend it to me?

Reply:
I think it's the trade-off. But how many events would be accumulated? Is it really a lot to be worried?

Reply:
Ok good I will test it

Reply:
It works!

Reply:
We made a conclusion not to test how events are accumulated and how much memory they used because we can assume it should be smaller than our thought.